how-to-sans-io.rst

```diff
@@ -0,0 +1,328 @@
+Writing I/O-Free (Sans-I/O) Protocol Implementations
+====================================================
+
+This informational document is an attempt to make the case for implementing
+network protocols without writing I/O, and to provide concrete assistance and
+instructions for doing so in Python.
+
+What Is An I/O-Free Protocol Implementation?
+--------------------------------------------
+
+An I/O-free protocol implementation (colloquially referred to as a "sans-IO"
+implementation) is an implementation of a network protocol that contains no
+code that does any form of network I/O or any form of asynchronous flow
+control. Put another way, a sans-IO protocol implementation is one that is
+defined entirely in terms of synchronous functions returning synchronous
+results, and that does not block or wait for any form of I/O. Examples of this
+kind of application can be found on :doc:`the landing page <index>`.
+
+Such a protocol implementation is, at the surface level, not very useful. After
+all, it's not very helpful to write a network protocol implementation that
+doesn't speak to the network! However, it turns out that writing protocol
+implementations in this manner provides a number of extremely useful benefits
+to the wider software ecosystem, as well as to the quality of your own code.
+
+In the rest of this document we'll outline what those benefits are, and also
+describe some of the techniques that are commonly employed to write protocol
+stacks in this manner.
+
+.. note:: It should be noted that the sans-IO implementation style is a
+          specific facet of several broader software design best practices. In
+          particular, good comparisons can be made to Bob Martin's `Clean
+          Architecture`_, to well-realised Model-View-Controller applications,
+          and to the broader software design principle of separation of
+          concerns. While all of these are good patterns worth being understood
+          in their own right, it is often helpful to discuss these patterns as
+          they apply to a specific problem domain: this is exactly what
+          sans-IO is all about.
+
+
+.. _why-bother:
+
+Why Write I/O-Free Protocol Implementations?
+--------------------------------------------
+
+Writing sans-IO protocol implementations provides a number of useful benefits,
+both to the implemenation itself and to the wider software ecosystem. We'll
+discuss these in turn.
+
+
+.. _simple-testable-correct:
+
+Simplicity, Testability, and Correctness
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Simplicity
+^^^^^^^^^^
+
+On a self-serving level, writing a protocol implementation containing no I/O
+makes writing a high-quality implementation substantially easier. It is no
+secret that network I/O is particularly prone to a wide variety of unexpected
+failure modes that can occur at almost any time, even in the simplest cases.
+When the protocol implementation no longer drives its own I/O, but instead has
```

Comment:
I would consider dropping the comma before "but" as there's another comma later and it makes it a little hard to read.